File: zozi/addition_array.rb
test_one = [4, 5, 9]
test_two = [1, 2, 3, 24]
test_three = [1, 4, 3, 8, 7, 2]
test_four = [1, 3, 4, 7, 2]

def addition(numbers)
  max = numbers.max
  numbers.delete(max)

  build_combinations(numbers) do |combo|
    combo.reduce(&:+) == max
  end
end

def build_combinations(rest)

  false
end

puts "test_one: #{test_one} returns \"#{addition(test_one)}\", expect it to be true"
#puts "test_two: #{test_two} returns \"#{addition(test_two)}\", expect it to be false"
#puts "test_three: #{test_three} returns \"#{addition(test_three)}\", expect it to be true"
#puts "test_four: #{test_four} returns \"#{addition(test_four)}\", expect it to be true"

File: zozi/addition_array.js
var expect = require('chai').expect;
var _ = require('underscore');

describe('additionArray', function () {
  it('returns true given [1, 2, 3]', function () {
    expect(additionArray([1, 2, 3])).to.eq(true);
  });

  it('returns false given [1, 2, 3, 10]', function () {
    expect(additionArray([1, 2, 3, 10])).to.eq(false);
  });

  it('returns true given [1, 2, 3, 10, 9]', function () {
    expect(additionArray([1, 2, 3, 10, 9])).to.eq(true);
  });

  it('returns true given [1, 20, 3, 10, 9]', function () {
    expect(additionArray([1, 20, 3, 10, 9])).to.eq(true);
  });
});

function additionArray(numbers) {
  var maxInt = _.max(numbers);
  var maxIdx = numbers.indexOf(maxInt);
  numbers.splice(maxIdx, 1);

  var combinations = buildCombination([], numbers, []);

  return !!_.find(combinations, function (combo) {
    return sum(combo) === maxInt;
  });
}

function buildCombination(active, rest, combinations) {
  if (!rest.length) {
    combinations.push(active);
  } else {
    buildCombination(active.concat(rest[0]), rest.slice(1), combinations);
    buildCombination(active, rest.slice(1), combinations);
  }

  return combinations;
}

function sum(numbers) {
  return _.reduce(numbers, function (sum, num) {
    return sum + num;
  }, 0);
}
